kidex-common/src/query.rs

```diff
@@ -0,0 +1,232 @@
+pub enum FileType {
+    All,
+    FilesOnly,
+    DirOnly,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct Keyword {
+    pub word: String,
+    pub exact_match: bool,
+}
+impl Keyword {
+    pub fn new(word: &str, exact_match: bool) -> Self {
+        Keyword {
+            exact_match,
+            word: word
+                .trim()
+                .trim_matches('/')
+                .to_string(),
+        }
+    }
+
+    pub fn is_at_beginning(&self, candidate: &str, case_options: &CaseOption) -> bool {
+        match case_options {
+            CaseOption::Match => candidate.starts_with(&self.word),
+            CaseOption::Ignore => candidate.to_lowercase().starts_with(&self.word.to_lowercase()),
+            CaseOption::Smart => {
+                if &self.word.to_lowercase() != &self.word {
+                    // Case sensitive
+                    candidate.starts_with(&self.word)
+                } else {
+                    // Ignoring case
+                    candidate.to_lowercase().starts_with(&self.word.to_lowercase())
+                }
+            }
+        }
+    }
+
+    pub fn is_in(&self, candidate: &str, case_options: &CaseOption) -> bool {
+        let (cased_candidate, cased_word) = match case_options {
+            CaseOption::Match => (candidate.into(), &self.word),
+            CaseOption::Ignore => (candidate.to_lowercase(), &self.word.to_lowercase()),
+            CaseOption::Smart => {
+                if &self.word.to_lowercase() != &self.word {
+                    // Case sensitive
+                    (candidate.into(), &self.word)
+                } else {
+                    // Ignoring case
+                    (candidate.to_lowercase(), &self.word.to_lowercase())
+                }
+            },
+        };
+        if self.exact_match {
+            cased_candidate == *cased_word
+        } else {
+            cased_candidate.contains(cased_word)
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct Query {
+    pub file_type: FileType,
+    pub direct_parent: Option<Keyword>,
+    pub keywords: Vec<Keyword>,
+    pub path_keywords: Vec<Keyword>,
+    pub case_option: CaseOption,
+}
+impl Default for Query {
+    fn default() -> Self {
+        Query {
+            file_type: FileType::All,
+            direct_parent: None,
+            keywords: vec![],
+            path_keywords: vec![],
+            case_option: CaseOption::Smart,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct QueryOptions {
+    pub query: Query,
+    pub output_format: OutputFormat,
+    pub root_path: Option<PathBuf>,
+    pub limit: Option<usize>,
+}
+impl Default for QueryOptions {
+    fn default() -> Self {
+        QueryOptions {
+            query: Query::default(),
+            output_format: OutputFormat::Json,
+            root_path: None,
+            limit: None,
+        }
+    }
+}
+
+impl Query {
```

Comment:
I think it would work better if every query was formed of a Vec of query parameters, and the query parameters would be parsed from a Vec of strings. This would allow the query parameters to be the same for any CLI tools, and possible runner integrations.